runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollOverlay.java

```diff
@@ -22,74 +23,321 @@
+					{
+						if (equippedItems.contains(itemId))
+						{
+							panelComponent.getLines().add(new PanelComponent.Line(itemDefinition.getName(), TITLED_CONTENT_COLOR, "X", Color.GREEN));
+						}
+						else
+						{
+							panelComponent.getLines().add(new PanelComponent.Line(itemDefinition.getName(), TITLED_CONTENT_COLOR, "-", Color.RED));
+						}
+					}
 				}
+			}
+			else
+			{
+				panelComponent.setWidth(130);
+				panelComponent.getLines().add(new PanelComponent.Line("Items:", "None"));
+			}
+		}
+		else if (clue instanceof FairyRingClue)
+		{
+			FairyRingClue fairyRingClue = (FairyRingClue) clue;
 
-				panelComponent.getLines().add(new PanelComponent.Line(clueItem.getName(), ""));
+			panelComponent.setTitle("Fairy Ring Clue");
+			panelComponent.setWidth(140);
+
+			panelComponent.getLines().add(new PanelComponent.Line("Code:", ""));
+			panelComponent.getLines().add(new PanelComponent.Line(fairyRingClue.getText().substring(0, 5), TITLED_CONTENT_COLOR, "", Color.WHITE));
+
+			panelComponent.getLines().add(new PanelComponent.Line("Travel to the fairy ring", ""));
+			panelComponent.getLines().add(new PanelComponent.Line("to see where to dig.", ""));
+		}
+		else if (clue instanceof MapClue)
+		{
+			MapClue mapClue = (MapClue) clue;
+
+			panelComponent.setTitle("Map Clue");
+
+			if (mapClue.isObjectClue())
+			{
+				panelComponent.setWidth(150);
+
+				ObjectComposition objectToClick = client.getObjectDefinition(mapClue.getObjectId());
+
+				String objectName = "N/A";
+				if (objectToClick != null)
+				{
+					objectName = objectToClick.getName();
+				}
+
+				panelComponent.getLines().add(new PanelComponent.Line("Travel to the destination", ""));
+				panelComponent.getLines().add(new PanelComponent.Line("and click the " + objectName + ".", ""));
+			}
+			else
+			{
+				panelComponent.setWidth(160);
+
+				panelComponent.getLines().add(new PanelComponent.Line("Travel to the destination", ""));
+				panelComponent.getLines().add(new PanelComponent.Line("and dig on the marked tile.", ""));
 			}
 		}
+		else if (clue instanceof WorldPoint)
+		{
+			panelComponent.setTitle("Coordinate Clue");
+			panelComponent.setWidth(135);
+
+			panelComponent.getLines().add(new PanelComponent.Line("Travel to the marked", ""));
+			panelComponent.getLines().add(new PanelComponent.Line("out destination to see", ""));
+			panelComponent.getLines().add(new PanelComponent.Line("a marker for where", ""));
+			panelComponent.getLines().add(new PanelComponent.Line("you should dig.", ""));
+		}
 
 		return panelComponent.render(graphics, parent);
 	}
+
+	private Set<Integer> getEquippedItems()
+	{
+		Set<Integer> equippedItems = new HashSet<>();
+
+		Item[] result = queryRunner.runQuery(new EquipmentItemQuery());
+
+		if (result != null)
+		{
+			for (Item item : result)
+			{
+				equippedItems.add(item.getId());
+			}
+		}
+
+		return equippedItems;
+	}
+
+	private List<String> lineBreakText(String text, int maxWidth, FontMetrics metrics)
```

Comment:
This seems pretty horrid here but offhand I'm not sure where this would go. @deathbeam ?

Reply:
I feel like this is something that TextComponent should do.